src/poisson_fft.f90

```diff
@@ -0,0 +1,198 @@
+module m_poisson_fft
+   use m_allocator, only: field_t
+   use m_common, only: dp, pi
+   use m_tdsops, only: dirps_t
+
+   implicit none
+
+   type, abstract :: poisson_fft_t
+      !! FFT based Poisson solver
+      !! It can only handle 1D decompositions along z direction.
+      integer :: nx, ny, nz
+      complex(dp), allocatable, dimension(:, :, :) :: waves
+      complex(dp), allocatable, dimension(:) :: ax, bx, ay, by, az, bz
+   contains
+      procedure(fft_forward), deferred :: fft_forward
+      procedure(fft_backward), deferred :: fft_backward
+      procedure(fft_postprocess), deferred :: fft_postprocess
+      procedure :: base_init
+      procedure :: waves_set
+   end type poisson_fft_t
+
+   abstract interface
+      subroutine fft_forward(self, f_in)
+         import :: poisson_fft_t
+         import :: field_t
+         implicit none
+
+         class(poisson_fft_t) :: self
+         class(field_t), intent(in) :: f_in
+      end subroutine fft_forward
+
+      subroutine fft_backward(self, f_out)
+         import :: poisson_fft_t
+         import :: field_t
+         implicit none
+
+         class(poisson_fft_t) :: self
+         class(field_t), intent(inout) :: f_out
+      end subroutine fft_backward
+
+      subroutine fft_postprocess(self)
+         import :: poisson_fft_t
+         implicit none
+
+         class(poisson_fft_t) :: self
+      end subroutine fft_postprocess
+   end interface
+
+contains
+
+   subroutine base_init(self, xdirps, ydirps, zdirps, sz)
+      implicit none
+
+      class(poisson_fft_t) :: self
+      class(dirps_t), intent(in) :: xdirps, ydirps, zdirps
+      integer, intent(in) :: sz
+
+      self%nx = xdirps%n; self%ny = ydirps%n; self%nz = zdirps%n
+
+      allocate (self%ax(self%nx), self%bx(self%nx))
+      allocate (self%ay(self%nx), self%by(self%nx))
+      allocate (self%az(self%nx), self%bz(self%nx))
+
+      allocate (self%waves(sz, self%nx, (self%ny*(self%nz/2 + 1))/sz))
+
+      ! waves_set requires some of the preprocessed tdsops variables.
+      call self%waves_set(xdirps, ydirps, zdirps, sz)
+
+   end subroutine base_init
+
+   subroutine waves_set(self, xdirps, ydirps, zdirps, sz)
```

Comment:
Unclear what this sub does. Needs a comment?

Reply:
To be honest, this is a black box for me too, I copied this from Incompact3D and there were no comments there either. 
@slaizet is probably the best person to add a description and few comments here if he has time and happy to do this, otherwise someone else who fully understands whats going on here, definitely not me. And this can be done in a separate PR not to delay this one any further. 
For now I'm only adding a description to reference the paper where its discussed in detail. 

https://github.com/xcompact3d/Incompact3d/blob/2da7699558806e70574a6980a75193a80f5734c9/src/poisson.f90#L1553

Reply:
I think best would be to seat together in the same room one afternoon and add as many comments as required (when at least one of us does not understand something basically). If we are all together, it should be quite quick and productive :-) We can discuss about this during our next weekly meeting on when best to add these comments in the new framework.

Reply:
Raised in #72